Metrics/AppMetrics.php

```diff
@@ -86,49 +93,69 @@ private function setInstance(string $value): void
         }
     }
 
-    private function incRequestsTotal(?string $method = null, ?string $route = null): void
+    private function incRequestsTotal(MetricInfo $metricInfo): void
     {
         $counter = $this->collectionRegistry->getOrRegisterCounter(
             $this->namespace,
             'http_requests_total',
             'total request count',
-            ['action']
+            $metricInfo->getLabelNames()
         );
 
-        $counter->inc(['all']);
+        $counter->inc($metricInfo->getLabelValueForAll());
 
-        if (null !== $method && null !== $route) {
-            $counter->inc([sprintf('%s-%s', $method, $route)]);
+        if ($this->isMetricInfoValid($metricInfo)) {
+            $counter->inc($metricInfo->getLabelValues());
         }
     }
 
-    private function incResponsesTotal(string $type, ?string $method = null, ?string $route = null): void
+    private function incResponsesTotal(string $type, MetricInfo $metricInfo): void
     {
         $counter = $this->collectionRegistry->getOrRegisterCounter(
             $this->namespace,
             sprintf('http_%s_responses_total', $type),
             sprintf('total %s response count', $type),
-            ['action']
+            $metricInfo->getLabelNames()
         );
-        $counter->inc(['all']);
 
-        if (null !== $method && null !== $route) {
-            $counter->inc([sprintf('%s-%s', $method, $route)]);
+        $counter->inc($metricInfo->getLabelValueForAll());
+
+        if ($this->isMetricInfoValid($metricInfo)) {
+            $counter->inc($metricInfo->getLabelValues());
         }
     }
 
-    private function setRequestDuration(float $duration, ?string $method = null, ?string $route = null): void
+    private function setRequestDuration(float $duration, MetricInfo $metricInfo): void
     {
         $histogram = $this->collectionRegistry->getOrRegisterHistogram(
             $this->namespace,
             'request_durations_histogram_seconds',
             'request durations in seconds',
-            ['action']
+            $metricInfo->getLabelNames()
         );
-        $histogram->observe($duration, ['all']);
 
-        if (null !== $method && null !== $route) {
-            $histogram->observe($duration, [sprintf('%s-%s', $method, $route)]);
+        $histogram->observe($duration, $metricInfo->getLabelValueForAll());
+
+        if ($this->isMetricInfoValid($metricInfo)) {
+            $histogram->observe($duration, $metricInfo->getLabelValues());
+        }
+    }
+
+    private function isMetricInfoValid(MetricInfo $metricInfo): bool
+    {
+        if (null !== $metricInfo->getRequestMethod() && null !== $metricInfo->getRequestRoute()) {
+            return true;
+        }
+
+        return false;
+    }
+
+    private function resolveMetricInfo(Request $request): MetricInfo
+    {
+        if (null === $this->metricInfoResolver) {
+            return new MetricInfo($request->getMethod(), $request->attributes->get('_route'));
```

Comment:
I like the idea to add custom labels. This should resolve #97 but you need to allow custom configuration of the additionnal labels in the `prometheus_metrics.yaml` first. Override the MetricInfo to enable this feature is a bit complex. 
It will be nice to have key-value in the configuration file which will be sent by extension or compiler in the AppMetrics.

Reply:
Gotcha, I will make those changes :+1:

Reply:
@Johnmeurt PR is updated, now it's possible to add custom labels more easily.